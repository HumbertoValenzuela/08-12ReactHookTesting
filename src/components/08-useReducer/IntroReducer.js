// 4._Idea_general_de_un_reducer_-_Vía_código
import React from 'react'

export const IntroReducer = () => {
    // Estado inicial un arreglo de "todos" objetos
    const initialState = [{
        id: 1,
        todo: 'Comprar pan',
        done: false
    }];

    // Reduce. Las action son la que modifica el state
    const todoReducer = ( state = initialState, action) => {
        // no puede realizar peticiones asincronas, no debe tener efecto secundario
        // Cannot read property 'type' of undefined
        // para corregir agregar ? action?
        if (action?.type === 'agregar') {
            return [ ...state, action.payload ];
        }
        // Retorna un nuevo estado
        return state;
    };    

    // Inicializacón del reducer
    let todos = todoReducer();
    console.log(todos);

    // Mandar newTodo a todoReducer
    const newTodo = {
        id: 2,
        todo: 'Comprar Leche',
        done: false
    };

    // Crear una acción para a agregar el newTodo.
    // crear un objeto con un standar. type le dice a reducer que tipo de acción es (agregar modificar etc).
    // payload es un standar de programación para trabajar con los argumentos de la action, la info viene en el payload
    const agregarTodoAction = {
        type: 'agregar',
        payload: newTodo
    }

    todos = todoReducer( todos, agregarTodoAction);
    console.log(todos);// muestra el Array de objeto

    return (
        <div>
            
        </div>
    )
}
