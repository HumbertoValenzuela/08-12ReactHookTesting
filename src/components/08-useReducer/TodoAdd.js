import React from 'react'
import { useForm } from '../../hooks/useForm';
// 12._Optimización_#2_-_Agregar_TODO
// Funciones y valores handleSubmit, handleInputChange, description se encuentran amarradas al useForm. ¿Cómo optimizar?
// Preguntar el componente TodoApp necesita que este useForm dentro de ella: respuesta es NO.
// Se agregará una handleAddTodo en el componenete TodoApp para realizar la comunicación
export const TodoAdd = ( { handleAddTodo } ) => {

    // Manejo de formulario
    const [ { description }, handleInputChange, reset ] = useForm( {  description: '' } );
    // console.log(description);
    // console.log(todos);

    const handleSubmit = ( e ) => {
        e.preventDefault();

        // Validar string vacios
        if( description.trim().length <= 1 ) {
            return;
        };

        const newTodo = {
            id: new Date().getTime(),           
            desc: description,
            done: false
        };

        // Utiliza la función que viene por parametro y envía el objeto newTodo
        handleAddTodo( newTodo );
      
        // reset de input
        reset();
    };

    return (
        <>
         <h4>Agregar TODO</h4>
 
            <form onSubmit={ handleSubmit }>
                <input 
                    autoComplete="off"
                    className="form-control"
                    name="description"
                    type="text"                     
                    placeholder="Aprender ..."     
                    onChange={handleInputChange} 
                    value={description} />                       
                <button
                    className="btn btn-outline-primary mt-1 btn-block"
                    type="submit">
                    Agregar
                </button>
            </form>           
        </>
    )
}
