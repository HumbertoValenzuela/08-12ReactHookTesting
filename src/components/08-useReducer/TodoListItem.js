import React from 'react'
 // TodoListItem, todo, index, handleDelete handleToggle
export const TodoListItem = ( { todo, indice, handleDelete, handleToggle } ) => {
    return (
        <li 
        key={todo.id}
        className="list-group-item"
        >
            <p 
                className={`${ todo.done && 'complete' }` }
                onClick= { () => handleToggle(todo.id)}
            >
                {indice + 1}. {todo.desc}
            </p> 
            <button 
                className="btn btn-danger"
                onClick={() => handleDelete(todo.id)}
            >
                borrar
            </button>
        </li>
    )
}
