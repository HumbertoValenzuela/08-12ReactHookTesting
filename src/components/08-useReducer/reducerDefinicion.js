// ¿Que es un Reducer?
// Es una Función. 
// Debe ser una función pura: No debe de tener efectos secundarios. No debe realizar tareas asíncronas. Debe de retornar un nuevo estado. No debe de llamar localStorage o sessionStorage internamente dentro de mi reducer, porque son efectos secundarios
//  Debe de retornar un nuevo estado
// Generalmente sólo recibe dos argumentos. El valor inicial (initialState) y la acción a ejecutar
// Tareas por hacer, las llamaremos "todos"
