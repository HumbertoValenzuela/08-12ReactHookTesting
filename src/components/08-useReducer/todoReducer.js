
export const todoReducer = ( state = [], action) => {
    // Se puede ocupar condicionales en esta sección
    switch ( action.type) {

        case 'add':
            // Obtener el state anterior y añadir el nuevo elemento
            return [ ...state, action.payload]
        case 'delete':
            // .filter regresa un new Array
            return state.filter(todo => todo.id !== action.payload);
        case 'toggle':
            return state.map(todo =>
                    ( todo.id === action.payload )
                        ? { ...todo, done: !todo.done }
                        : todo
                )
        // case 'toggle-old':            
        //     return state.map(todo => { 
        //         if ( todo.id === action.payload ) {
        //             // retornar todo los elementos pero se quiere cambiar el done
        //             return {
        //                 ...todo,
        //                 done: !todo.done
        //             }
        //         } else {
        //             // Sino cambia regresar algo, sino da error undefined
        //             return todo;
        //         }
               
        //     });
        default:
            return state;
    }
}