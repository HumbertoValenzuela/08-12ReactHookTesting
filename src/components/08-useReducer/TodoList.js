import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (        
        <ul className="list-group list-group-flush">
         {
             todos.map( (todo, indice) => (                
                 <TodoListItem
                 key= { todo.id }
                 todo={todo}
                 indice= {indice}
                 handleDelete={handleDelete}
                 handleToggle={handleToggle}
                 />                
             ))
         }     
        </ul>
    )
}
