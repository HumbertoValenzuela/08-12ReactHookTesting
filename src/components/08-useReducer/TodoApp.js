import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
import './styles.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

// reducer es la función reducer,
    // initializerArg es el estado inicial que se puede mandar
    // init es usado como una función ára inicializar el initializerArg, en caso de que sea como procesado o haga varias accciones.
    // init: cada vez que el componente se redibuja , entonces al ser una función  que esta externa al compoenente lo hace más ligero y tambien puede memorizar
    // state
    // dispatch: despacha las acciones asía mi reducer. Cuando se hace cambios en el initializerArg, React se enterá para redibujar en pantalla

   

    // Punto inicial de la aplicación
// const initializerArg = [{
//     id: new Date().getTime(),
//     desc: 'Aprender React',
//     done: false
// }];

// initializerArg queda dentro de la función init , pero se realiza un return del objeto. initializerArg en el useReducer queda []
const init = () => {

    // Extraer de LocalStorage
    // localStorage.getItem('noexiste') es null
    // JSON.parse(null) es null
    return JSON.parse(localStorage.getItem('todos') ) || [];
     //Extrae y si es null muestra []

    // return [{
    //         id: new Date().getTime(),
    //         desc: 'Aprender React',
    //         done: false
    //     }];
}

export const TodoApp = () => {

    
    // dispatch: función con una acción, sabe a que reducer mandar info
    // y cuando cambia initializerArg va a redibujar lo que cambie
    // init, ayudará a computar el estado inicial para que funcione más rápido el componente, y se re vuelva a ejecutar
    const [ todos, dispatch ] = useReducer(todoReducer, [], init );

    // Custom hook maneja el formulario
    // const [ formValues, handleInputChange] = useForm( {  description: '' });// Se asocia con el input name= description
    // console.log(formValues);
   
    // initializer

    // Grabar en LocalStorage. ¿Cuando? cuando los 'todos' cambien. usar useEffect
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
        // return () => {
        //     cleanup
        // }
    }, [todos]);
    // 10._Toggle_Todo_-_ Marcar_como_completado _ o _ pendiente_un_TODO
    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    };

    // Se puede solo la descripcion o todo el newTodo (objeto)
    const handleAddTodo = ( newTodo ) => {
        // ¿Como mandar action al todoReducer? con dispatch          
        dispatch( { 
            type: 'add',
            payload: newTodo
        } );
    };

    const handleDelete = (todoId) => {
        console.log(todoId);
        // crear la action borrar
        const action = {
            type: 'delete',
            payload: todoId
        }
        // dispatch
        dispatch(action);
    }   

    // 6._Creando_el_cascaron_de_la_lista_de_TODOs
    return (
        <div>
             <h2>TodoApp ( { todos.length } )</h2>

            <div className="row">

                <div className="col-7">
                <TodoList 
                    todos={todos} 
                    handleDelete={handleDelete}
                    handleToggle={handleToggle}
                    />
                </div>

                <div className="col-5">
                   <TodoAdd 
                //    Mandar por referencia a TodoAdd
                    handleAddTodo={ handleAddTodo } />
                </div>
            </div>
        </div>
    )
}
