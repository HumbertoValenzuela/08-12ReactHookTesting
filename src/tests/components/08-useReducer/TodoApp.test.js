import React from 'react';
import { shallow, mount } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../datosFijos/demoTodos';
import { act } from '@testing-library/react';

describe('Pruebas en <TodoApp />', () => {
    
    const wrapper = shallow( <TodoApp /> );

    // Verificar si llamamos el localStorage.
    Storage.prototype.setItem = jest.fn( ()=> {});

    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();  
    });

    // Llamar métodos como el handleAddTodo={ handleAddTodo } 
    test('debe de agregar un TODO', () => {
        // la dif es que no hará match con el shallow. Se necesita montar todo porque se quiere ver el localStorage, simular
        // mount de enzyme: Se ocupa para probar ya toda la aplicación en contexto, en general. Funciona igual que shallow la diferencia, es el nivel el cuál es renderizada la aplicación
        const wrapper = mount( <TodoApp /> );

        // Buscar elemento TodoAdd. Este componente se agrega una prop que se llama handleAddTodo
        //  Warning: An update to TodoApp inside a test was not wrapped in act(...).
        //  Se debe agregar el act
        // wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] );
        // Como se a montado y se a realizado un cambio. Se necesita envolverlo en el act
        act( () => {
            wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] );
            wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[1] );
        });

        // Se realizaron dos cambios, se agregaron
        // Se espera que el h1 tenga un texto TodoApp ( 2 )

        // console.log(wrapper.debug()) cuidado que si se repite el h2 dará error en .text que debe haber uno
        
        expect( wrapper.find('h2').text() ).toBe('TodoApp ( 2 )');

        // Revisar el useEffect(() => {localStorage.setItem('todos'
        // Verificar si llamamos el localStorage. Verificar que el setItem se a llamado. Pero no que se grabo o elimino
        expect( localStorage.setItem).toHaveBeenCalledTimes(2);
        // Para revisar lo que tiene dará error
        // expect( localStorage.setItem).toHaveBeenCalledWith({});
    });

    test('debe de eliminar un todo', () => {
        // Agrega uno
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        // Entra a todoList y elimina uno
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

        expect( wrapper.find('h2').text().trim() ).toBe('TodoApp ( 0 )');
    })
    
    
})
