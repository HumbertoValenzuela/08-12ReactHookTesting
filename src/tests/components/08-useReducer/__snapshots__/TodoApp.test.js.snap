// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h2>
    TodoApp ( 
    0
     )
  </h2>
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
